[libpas] Fix pas_segregated_directory_get_use_epoch

Constellation opened the pull request:
#### 4a099b3ac61b95b0e8304c6d6b6d1dea93827f7f
<pre>
[libpas] Fix pas_segregated_directory_get_use_epoch
<a href="https://bugs.webkit.org/show_bug.cgi?id=242911">https://bugs.webkit.org/show_bug.cgi?id=242911</a>
rdar://87308366

Reviewed by Mark Lam.

We observed some crashes in pas_segregated_directory_get_use_epoch assertion which is saying num_non_empty_words and use_epoch are both zero.
This can happen since pas_segregated_directory_get_use_epoch is not taking the page lock. We are taking the ownership lock, but this does not
guarantee that num_non_empty_words and use_epoch will not be modified. However, the problem is that (1) taking the ownership lock is required
to keep the page alive, and (2) taking both the page lock and the owership lock is not possible because of lack of locking ordering between them
throughout the code.

To fix this issue, we apply two changes.
Keep in mind that this pas_segregated_directory_get_use_epoch returns approximate value. If it is not precise in the face of race condition, it is OK.

1. Whenever changing num_non_empty_words to zero, we update use_epoch atomically. And in pas_segregated_directory_get_use_epoch, we load both
   atomically. This ensures that, whenever num_non_empty_words is zero, we do not get a stale value for use_epoch. This fixes one issue: we had
   small race window, after num_non_empty_words becomes zero and resetting use_epoch a new epoch, old use_epoch value can be read from the other
   thread while num_non_empty_words is zero. By doing this, we can ensure that old use_epoch value will not be visible to the other thread if
   num_non_empty_words is zero. This is one source of this crash because the newly created fresh page&apos;s use_epoch is zero. If the page is new and
   we encounter this race condition, we will get zero num_non_empty_words and use_epoch.
2. Still it is possible that we observe zero num_non_empty_words and use_epoch. This is because this function is not taking eligible bit at all.
   This means that, between checking empty bit and taking the onwership lock, the underlying page can be purged, reallocated, or whatever can happen.
   If the page is destroyed and reallocated, and just before actually starting use of this page, this page&apos;s num_non_empty_words and use_epoch are both
   zero. In this case, we can just ignore this page since it is already in use.

* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_scan_bits_to_set_up_free_bits):
(pas_local_allocator_prepare_to_allocate):
(pas_local_allocator_set_up_primordial_bump):
* Source/bmalloc/libpas/src/libpas/pas_page_base.c:
(pas_page_base_is_empty):
* Source/bmalloc/libpas/src/libpas/pas_segregated_directory.c:
(pas_segregated_directory_get_use_epoch):
* Source/bmalloc/libpas/src/libpas/pas_segregated_page.c:
(pas_segregated_page_construct):
(pas_segregated_page_note_emptiness):
(pas_segregated_page_take_physically):
* Source/bmalloc/libpas/src/libpas/pas_segregated_page.h:
(pas_segregated_page_qualifies_for_decommit):
* Source/bmalloc/libpas/src/libpas/pas_segregated_page_inlines.h:
(pas_segregated_page_deallocate_with_page):
* Source/bmalloc/libpas/src/libpas/pas_segregated_partial_view.c:
(pas_segregated_partial_view_should_table):
* Source/bmalloc/libpas/src/libpas/pas_segregated_shared_page_directory.c:
(take_last_empty_consider_view):
* Source/bmalloc/libpas/src/libpas/pas_segregated_size_directory.c:
(take_last_empty_consider_view):
* Source/bmalloc/libpas/src/libpas/pas_utils.h:
(pas_atomic_store_pair_relaxed):

Canonical link: <a href="https://commits.webkit.org/252635@main">https://commits.webkit.org/252635@main</a>
</pre>

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_segregated_page.h:
```diff
@@ -74,7 +83,7 @@ struct pas_segregated_page {
     pas_lock* lock_ptr;
 
-    uintptr_t use_epoch;
+    struct pas_segregated_page_emptiness emptiness;
```
nit: why not typedef the struct above as we already do for other structs, and forego having to explicitly declaring it as a struct here and everywhere we use `pas_segregated_page_emptiness`?

Constellation replied:
No particular reason. Changed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_segregated_page.h:
```diff
@@ -413,7 +421,12 @@ pas_segregated_page_is_allocated(uintptr_t begin,
+typedef enum {
+    pas_note_emptiness_clear_num_non_empty_words,
+    pas_note_emptiness_keep_num_non_empty_words,
+} pas_note_emptiness_mode;
```
Let's call this `pas_note_emptiness_action` instead as it indicates the action we want.

Constellation replied:
Sounds good, fixed.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[libpas] Fix pas_segregated_directory_get_use_epoch
https://bugs.webkit.org/show_bug.cgi?id=242911
rdar://87308366

Reviewed by Mark Lam.

We observed some crashes in pas_segregated_directory_get_use_epoch assertion which is saying num_non_empty_words and use_epoch are both zero.
This can happen since pas_segregated_directory_get_use_epoch is not taking the page lock. We are taking the ownership lock, but this does not
guarantee that num_non_empty_words and use_epoch will not be modified. However, the problem is that (1) taking the ownership lock is required
to keep the page alive, and (2) taking both the page lock and the owership lock is not possible because of lack of locking ordering between them
throughout the code.

To fix this issue, we apply two changes.
Keep in mind that this pas_segregated_directory_get_use_epoch returns approximate value. If it is not precise in the face of race condition, it is OK.

1. Whenever changing num_non_empty_words to zero, we update use_epoch atomically. And in pas_segregated_directory_get_use_epoch, we load both
   atomically. This ensures that, whenever num_non_empty_words is zero, we do not get a stale value for use_epoch. This fixes one issue: we had
   small race window, after num_non_empty_words becomes zero and resetting use_epoch a new epoch, old use_epoch value can be read from the other
   thread while num_non_empty_words is zero. By doing this, we can ensure that old use_epoch value will not be visible to the other thread if
   num_non_empty_words is zero. This is one source of this crash because the newly created fresh page's use_epoch is zero. If the page is new and
   we encounter this race condition, we will get zero num_non_empty_words and use_epoch.
2. Still it is possible that we observe zero num_non_empty_words and use_epoch. This is because this function is not taking eligible bit at all.
   This means that, between checking empty bit and taking the onwership lock, the underlying page can be purged, reallocated, or whatever can happen.
   If the page is destroyed and reallocated, and just before actually starting use of this page, this page's num_non_empty_words and use_epoch are both
   zero. In this case, we can just ignore this page since it is already in use.

* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_scan_bits_to_set_up_free_bits):
(pas_local_allocator_prepare_to_allocate):
(pas_local_allocator_set_up_primordial_bump):
* Source/bmalloc/libpas/src/libpas/pas_page_base.c:
(pas_page_base_is_empty):
* Source/bmalloc/libpas/src/libpas/pas_segregated_directory.c:
(pas_segregated_directory_get_use_epoch):
* Source/bmalloc/libpas/src/libpas/pas_segregated_page.c:
(pas_segregated_page_construct):
(pas_segregated_page_note_emptiness):
(pas_segregated_page_take_physically):
* Source/bmalloc/libpas/src/libpas/pas_segregated_page.h:
(pas_segregated_page_qualifies_for_decommit):
* Source/bmalloc/libpas/src/libpas/pas_segregated_page_inlines.h:
(pas_segregated_page_deallocate_with_page):
* Source/bmalloc/libpas/src/libpas/pas_segregated_partial_view.c:
(pas_segregated_partial_view_should_table):
* Source/bmalloc/libpas/src/libpas/pas_segregated_shared_page_directory.c:
(take_last_empty_consider_view):
* Source/bmalloc/libpas/src/libpas/pas_segregated_size_directory.c:
(take_last_empty_consider_view):
* Source/bmalloc/libpas/src/libpas/pas_utils.h:
(pas_atomic_store_pair_relaxed):

Canonical link: https://commits.webkit.org/252635@main
